data_structures/rb_tree2.cpp

```diff
@@ -0,0 +1,1728 @@
+/**
+ * @file
+ * @author [Brennan Allen](https://github.com/brennandoubt)
+ * (brennandoubt@gmail.com)
+ * @brief A Red-Black Tree (RBT) program with test cases.
+ * @version 0.1
+ * @date 2022-12-29
+ * 
+ * Copyright (no copyright)
+ * 
+ * @details A program simulating the Red Black Tree algorithm
+ * using helper functions with descriptive function names and annotations.
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <sstream>   /// for creating ostringstreams
+#include <string>    /// for creating strings
+#include <vector>    /// for creating vectors
+
+using std::cout;
+using std::endl;
+using std::invalid_argument;
+using std::ostringstream;
+using std::string;
+using std::vector;
+
+/// defining colors (const instead of defined macros)
+const int COLOR_RED = 0;
+const int COLOR_BLACK = 1;
+const int COLOR_DOUBLE_BLACK = 2;
+
+/**
+ * @brief Node struct for nodes inside the
+ * red-black tree. Each node contains data for its color
+ * (represented as a number), an integer value, and pointers to its
+ * parent node and left and right child nodes.
+ *
+ */
+struct RBTNode {
+    /// nodes are red by default
+    /// null/root/red's child nodes are recolored to black
+    uint16_t color = COLOR_RED;
+    /// has int value to represent node itself (default 0)
+    int val = 0;
+
+    /// has a parent and left/right child nodes
```

Comment:
these doc comments should probably be either double slashes or moved into the @details section of the struct